lib/rspec/buildkite/analytics/session.rb

```diff
@@ -177,11 +172,53 @@ def connect
       wait_for_confirm
 
       @logger.write("connected")
+
+      # As this connect method can be called multiple times in the
+      # reconnection process, kill prev write threads (if any) before
+      # setting up the new one
+      @write_thread&.kill
+
+      @write_thread = Thread.new do
+        @logger.write("hello from write thread")
+        # Pretty sure this while loop is fine cos the call to queue.pop is blocking
+        while
+          data = @send_queue.pop
+          message_type = data["action"]
+
+          if message_type == "end_of_transmission"
+            # Because of the unpredictable sequencing between the test suite finishing
+            # (EOT gets queued) and disconnections happening (retransmit results gets
+            # queued), we don't want to send an EOT before any retransmits are sent.
+            unless @send_queue.empty?
+              @send_queue << data
+              @logger.write("putting eot at back of queue")
+              next
+            end
+            @eot_queued_mutex.synchronize do
+              @eot_queued = false
+            end
+          end
+
+          @connection.transmit({
+            "identifier" => @channel,
+            "command" => "message",
+            "data" => data.to_json
```

Comment:
`results` in this value has already had `to_json` called on it on line 247. This results in a json object that gets reencoded as a json escaped string. Might this have something to do with your PR to bk/bk?